src/zjs_callbacks.c

```diff
@@ -387,64 +386,86 @@ void print_callbacks(void)
 #define print_callbacks() do {} while (0)
 #endif
 
-void zjs_call_callback(int32_t i)
+void zjs_call_callback(int32_t id, void* data, uint32_t sz)
 {
-    if (cb_map[i]->type == CALLBACK_TYPE_JS) {
-        if (cb_map[i]->js->func_list == NULL && jerry_value_is_function(cb_map[i]->js->js_func)) {
-            uint32_t argc = 0;
-            jerry_value_t ret_val;
-            jerry_value_t* args = NULL;
-
-            if (cb_map[i]->js->pre) {
-                args = cb_map[i]->js->pre(cb_map[i]->js->handle, &argc);
-            }
-
-            DBG_PRINT("calling callback id %ld with %lu args\n", cb_map[i]->js->id, argc);
-            ret_val = jerry_call_function(cb_map[i]->js->js_func, cb_map[i]->js->this, args, argc);
-            if (jerry_value_has_error_flag(ret_val)) {
-                DBG_PRINT("callback %ld returned an error\n", i);
-            }
-            if (cb_map[i]->js->post) {
-                cb_map[i]->js->post(cb_map[i]->js->handle, &ret_val);
-            }
-            if (cb_map[i]->js->once) {
-                zjs_remove_callback(i);
-            }
-        } else {
-            int j;
-            uint32_t argc = 0;
-            jerry_value_t ret_val;
-            jerry_value_t* args = NULL;
-
-            if (cb_map[i]->js->pre) {
-                args = cb_map[i]->js->pre(cb_map[i]->js->handle, &argc);
-            }
-
-            DBG_PRINT("calling callback list id %ld with %lu args\n", cb_map[i]->js->id, argc);
+    if(cb_map[id]) {
+        if (cb_map[id]->type == CALLBACK_TYPE_JS) {
+            if (cb_map[id]->js->func_list == NULL && jerry_value_is_function(cb_map[id]->js->js_func)) {
+                jerry_value_t ret_val;
+                // Single function callback
+                if (sz) {
+                    ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, data, sz);
+                } else {
+                    ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, NULL, 0);
+                }
```

Comment:
Should be able to drop this else clause. They should be passing data == NULL already anyway right? And if you pass size 0 it shouldn't read data anyway. (BTW you might want to update zjs_callbacks.h jsdoc to define the data/size parameter for this function and there you could be explicit that you pass NULL if it's 0 size, if you want.